external_search.py

```diff
@@ -882,8 +577,745 @@ def create_new_version(cls, search_client, base_index_name, version=None):
+
+    def add_match_query(self, query, field, query_string, matched_portion):
+        """Create a match query that finds documents whose value for `field`
+        matches `query`.
+
+        Add it to `self.match_queries`, and remove the relevant portion
+        of `query_string` so it doesn't get reused.
+        """
+        if not query:
+            # This is not a relevant part of the query string.
+            return query_string
+        match_query = self.query_class._match(field, query)
+        self.match_queries.append(match_query)
+        return self._without_match(query_string, matched_portion)
+
+    def add_target_age_query(self, query, query_string, matched_portion):
+        """Create a query that finds documents whose value for `target_age`
+        matches `query`.
+
+        Add it to `match_queries`, and remove the relevant portion
+        of `query_string` so it doesn't get reused.
+        """
+        if not query:
+            # This is not a relevant part of the query string.
+            return query_string
+        match_query = self.query_class.make_target_age_query(query, 40)
+        self.match_queries.append(match_query)
+        return self._without_match(query_string, matched_portion)
+
+    @classmethod
+    def _without_match(cls, query_string, match):
+        """Take the portion of a query string that matched a controlled
+        vocabulary, and remove it from the query string, so it
+        doesn't get reused later.
+        """
+        # If the match was "children" and the query string was
+        # "children's", we want to remove the "'s" as well as
+        # the match. We want to remove everything up to the
+        # next word boundary that's not an apostrophe or a
+        # dash.
+        word_boundary_pattern = r"\b%s[\w'\-]*\b"
+
+        return re.compile(
+            word_boundary_pattern % match.strip(), re.IGNORECASE
+        ).sub("", query_string)
+
+
+class Filter(SearchBase):
+    """A filter for search results.
+
+    This covers every reason you might want to not show a search
+    result that matches the query string -- wrong media, wrong
+    language, not available in the patron's library, etc.
+    """
+
+    @classmethod
+    def from_worklist(cls, _db, worklist, facets):
+        """Create a Filter that finds only works that belong in the given
+        WorkList and EntryPoint.
+
+        :param worklist: A WorkList
+        :param facets: A SearchFacets object.
+        """
+        library = worklist.get_library(_db)
+
+        v = worklist.inherited_value
+        media = v('media')
+        languages = v('languages')
+        fiction = v('fiction')
+        audiences = v('audiences')
+        target_age = v('target_age')
+
+        # For genre IDs and CustomList IDs, we might get a separate
+        # set of restrictions from every item in the WorkList hierarchy.
+        # _All_ restrictions must be met for a work to match the filter.
```

Comment:
This is the new bit. The filter may end up with multiple sets of restrictions on `genre_ids` or `customlist_ids`, and the way filters work, every restriction must be met for the book to show up. My goal is to use this to replace the multiple aliased joins against `Genre` and `CustomListEntry` in `WorkList.bibliographic_filter_clause` and `WorkList.customlist_filter_clause`.